app/src/main/java/com/github/se/cyrcle/model/map/BikeLocationViewModel.kt

```diff
@@ -0,0 +1,177 @@
+     */
+    fun checkValidity(): Boolean {
+      if (!isValid) return false
+      if (location == null) return false
+      return !location.longitude().isNaN() and !location.latitude().isNaN()
+    }
+
+    companion object {
+      /** A constant to represent an invalid location. */
+      val INVALID_LOCATION = BikeLocation(false, null)
+      /** A constant to represent an invalid double value for longitude and/or latitude. */
+      const val INVALID_DOUBLE = Double.NaN
+    }
+  }
+
+  private val _bikeLocation = MutableStateFlow(INVALID_LOCATION)
+  val bikeLocation = _bikeLocation
+
+  init {
+    Log.d("BikeLocationViewModel", "Init | Fetching bike location from Preferences")
+
+    runBlocking {
+      val bikeLocation =
+          BikeLocation(
+              preferenceStorage.readPreference(isValidPreferenceKey, false),
+              preferenceStorage.readPreference(longitudePreferenceKey, INVALID_DOUBLE),
+              preferenceStorage.readPreference(latitudePreferenceKey, INVALID_DOUBLE),
+          )
+
+      if (bikeLocation.checkValidity()) {
+        Log.d("BikeLocationViewModel", "Bike location is valid : $bikeLocation")
+        _bikeLocation.update { bikeLocation }
+        _bikeLocationState.update { LOCATION_STORED_UNUSED }
+      } else {
+        Log.d("BikeLocationViewModel", "Bike location is invalid : $bikeLocation")
+        _bikeLocation.update { INVALID_LOCATION }
+        _bikeLocationState.update { NO_LOCATION }
+      }
+    }
+  }
+
+  /**
+   * Save persistently the [location] of a bike. Can be called in any state, and will update the
+   * state to [LOCATION_STORED_UNUSED].
+   *
+   * @param location The location at which the bike is.
+   */
+  fun parkBike(location: Point) {
+    _bikeLocation.update { BikeLocation(true, location) }
+    _bikeLocationState.update { LOCATION_STORED_UNUSED }
+
+    runBlocking {
+      preferenceStorage.writePreference(isValidPreferenceKey, _bikeLocation.value.isValid)
+      preferenceStorage.writePreference(
+          longitudePreferenceKey, _bikeLocation.value.location!!.longitude())
+      preferenceStorage.writePreference(
+          latitudePreferenceKey, _bikeLocation.value.location!!.latitude())
+    }
+  }
+
+  /**
+   * Fetch the location of the bike from the preferences and update the state of the bike location.
+   * Will only have effect if the current state is [LOCATION_STORED_USED] or
+   * [LOCATION_STORED_UNUSED]. If the fetched location is invalid (i.e. IO errors or data
+   * corruption), the state will be updated to [NO_LOCATION] and the bike location will be set to
+   * [INVALID_LOCATION].
+   *
+   * @return The location of the bike. Will be null if the view model is not in a correct state. It
+   *   will be a valid location only if there was no IO error neither data corruption.
+   */
+  fun goToMyBike(): BikeLocation? {
+    Log.d("BikeLocationViewModel", "Fetching bike location from Preferences")
```

Comment:
You left some Log, are they for debugging purpose or did you forget to remove them ?